File: source/Artemis-VM/std_lib/KernelInterrupts.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace Artemis_IL.StandardLib
{
    /// <summary>
    /// Implements the AIL kernel interrupt handler (KEI instruction, opcode 0x2B).
    ///
    /// <b>Usage in AIL source:</b>
    /// <code>
    ///   KEI 0x01   ; stdio I/O — behaviour selected by register AL
    ///   KEI 0x02   ; halt the VM
    /// </code>
    ///
    /// <b>KEI 0x01 — stdio (behaviour controlled by AL):</b>
    /// <list type="table">
    ///   <listheader><term>AL</term><description>Action</description></listheader>
    ///   <item><term>0x01</term><description>Write the character in AH to stdout.</description></item>
    ///   <item><term>0x02</term><description>Write a string from RAM: X = base address, B = byte count.</description></item>
    ///   <item><term>0x03</term><description>Read one character from stdin into AH.</description></item>
    ///   <item><term>0x04</term><description>Read a line from stdin into RAM at address X; B is set to the number of bytes written.</description></item>
    ///   <item><term>0x05</term><description>Write the 16-bit value of register B as a decimal integer to stdout.</description></item>
    /// </list>
    ///
    /// <b>KEI 0x02 — halt:</b>
    /// Writes "Halting!" to stdout and stops VM execution.
    /// </summary>
    public static class KernelInterrupts
    {
        /// <summary>The VM instance that issued the KEI instruction.</summary>
        public static VM ParentVM;

        /// <summary>
        /// Dispatches a kernel interrupt.
        /// </summary>
        /// <param name="command">The interrupt number taken from the KEI operand (byte 1 of the instruction).</param>
        public static void HandleInterrupt(int command)
        {
            // ── KEI 0x01 — stdio ─────────────────────────────────────────────
            #region stdio
            if (command == 0x01)
            {
                if (Globals.DebugMode == true)
                {
                    Globals.console.Write("KEI 0x01: ");
                }

                if (ParentVM.AL == 0x01)
                {
                    // Write-character mode: output the single byte stored in AH as a char.
                    Globals.console.Write((char)ParentVM.AH);
                    if (Globals.DebugMode == true)
                        Globals.console.WriteLine(" 0x01");
                }
                else if (ParentVM.AL == 0x02)
                {
                    // Write-string mode: read B bytes from RAM starting at address X,
                    // convert each byte to a character and write them all to stdout.
                    byte[] toConvert = new byte[ParentVM.GetSplit('B')];
                    toConvert = ParentVM.ram.GetSection(ParentVM.X, ParentVM.GetSplit('B'));
                    string toPrint = "";
                    for (int i = 0; i < toConvert.Length; i++)
                    {
                        toPrint += (char)toConvert[i];
                    }
                    if (Globals.DebugMode == true)
                    {
                        Globals.console.WriteLine(" 0x02");
                    }
                    Globals.console.Write(toPrint);
                }
                else if (ParentVM.AL == 0x03)
                {
                    // Read-character mode: read one character from stdin and store it in AH.
                    ParentVM.AH = (byte)Globals.console.Read();
                }
                else if (ParentVM.AL == 0x04)
                {
                    // Read-line mode: read a line of text from stdin, store the raw bytes in
                    // RAM at address X, and set register B to the number of bytes written.
                    string toConvert = Globals.console.ReadLine();
                    byte[] toWrite = new byte[toConvert.Length];
                    for (int i = 0; i < toWrite.Length; i++)
                    {
                        toWrite[i] = (byte)toConvert[i];
                    }
                    ParentVM.SetSplit('B', toWrite.Length);
                    ParentVM.ram.SetSection(ParentVM.X, toWrite);
                }
                else if (ParentVM.AL == 0x05)
                {
                    // Write-integer mode: print the 16-bit value in register B as a
                    // decimal (base-10) ASCII string.  Useful for printing numeric results
                    // without manually converting digits.
                    Globals.console.Write(ParentVM.GetSplit('B').ToString());
                }
            }
            // ── KEI 0x02 — halt ──────────────────────────────────────────────
            else if (command == 0x02)
            {
                if (Globals.DebugMode == true)
                {
                    Globals.console.Write("KEI 0x02: ");
                }
                // Print the standard halt message and stop the execution loop.
                Globals.console.WriteLine("Halting!");
                ParentVM.Halt();
            }
            else
            {
                // Unknown interrupt number — halt to protect program state.
                Globals.console.WriteLine("Undocumented function: " + command + "\nHalting for protection of data");
                ParentVM.Halt();
            }
            #endregion
        }
    }
}

File: source/AIL-CPP/src/vm/kernel_interrupts.cpp
#include "ail/vm/kernel_interrupts.hpp"
#include "ail/vm/vm.hpp"
#include "ail/config.hpp"
#include <cstdio> // EOF

namespace ail {

void KernelInterrupts::handleInterrupt(VM& vm, int command) {
    switch (command) {

    // ── KEI 0x01 — stdio ─────────────────────────────────────────────────────
    case 0x01:
        switch (vm.AL) {

        case 0x01: // Write the single character in AH to stdout
            AIL_PUTCHAR(static_cast<char>(vm.AH));
            break;

        case 0x02: { // Write B bytes from RAM at address X to stdout
            int  count = vm.getSplit('B');
            int  base  = static_cast<int>(vm.X);
            for (int i = 0; i < count; ++i)
                AIL_PUTCHAR(static_cast<char>(vm.ram.getByte(base + i)));
            break;
        }

        case 0x03: // Read one character from stdin into AH
            vm.AH = static_cast<uint8_t>(AIL_GETCHAR());
            break;

        case 0x04: { // Read a line from stdin into RAM at address X; B = length
            int base = static_cast<int>(vm.X);
            int n    = 0;
            int ch;
            while ((ch = AIL_GETCHAR()) != EOF && ch != '\n') {
                vm.ram.setSection(base + n, reinterpret_cast<const uint8_t*>("\0"), 0);
                // Write byte directly (within data area — setSection bypasses guard)
                if (base + n < RAM::SIZE)
                    vm.ram.memory[base + n] = static_cast<uint8_t>(ch);
                ++n;
            }
            vm.setSplit('B', n);
            break;
        }

        case 0x05: // Write the 16-bit value of B as decimal
            AIL_PRINTF("%d", vm.getSplit('B'));
            break;

        default:
            break;
        }
        break;

    // ── KEI 0x02 — halt ───────────────────────────────────────────────────────
    case 0x02:
        AIL_PUTS("Halting!\n");
        vm.halt();
        break;

    default:
        AIL_PRINTF("Undocumented KEI 0x%02X — halting for data protection\n", command);
        vm.halt();
        break;
    }
}

} // namespace ail
